### DISCUSSION TITLE
Deadbeef plays all music files at the wrong pitch/speed

### DISCUSSION THREAD

**Discussion by thanks6** [DeaDBeeF-Player/deadbeef#2680] [General]
### Information about the software:

Deadbeef version: 1.8.2
OS: Linux Mint Kernel: 5.3.0-62-generic x86_64 bits: 64 compiler: gcc v: 7.5.0 Desktop: Cinnamon 4.4.8 
Distro: Linux Mint 19.3 Tricia base: Ubuntu 18.04 bionic

Hi
I am relatively new to Linux so I would appreciate any help/advice with regards to getting the best audio from my rather old system. I want to be able to play PCM as flac and DSD (although that is probably a separate issue). My flac files are both red book 16/44.1 and higher res up to 24/192. I have a USB DAC (Ifi Zen Dac) which is recognised by the OS. The playback issues occur with both the on-board audio card and the external USB DAC. 

Deadbeef can only play music files correctly with Pulse. With Alsa it plays all types of music files (flac, wav, Mp3) at the wrong pitch - 16/44.1 plays a little too fast and higher res eg 24/96 very slow. My aim is to get Deadbeef to work perfectly with Alsa to obtain a bit-perfect playback system. Any help greatly appreciated. Thank you

System: Host: GA-MA790FXT-UD5P Kernel: 5.3.0-62-generic x86_64 bits: 64 compiler: gcc v: 7.5.0 Desktop: Cinnamon 4.4.8 
Distro: Linux Mint 19.3 Tricia base: Ubuntu 18.04 bionic 
Machine: Type: Desktop Mobo: Gigabyte model: GA-MA790FXT-UD5P v: x.x serial: <filter> BIOS: Award v: F8n date: 01/07/2011 
CPU: Topology: Quad Core model: AMD Phenom II X4 955 bits: 64 type: MCP arch: K10 rev: 2 L2 cache: 2048 KiB 
flags: lm nx pae sse sse2 sse3 sse4a svm bogomips: 25716 
Speed: 3200 MHz min/max: 800/3200 MHz Core speeds (MHz): 1: 3200 2: 800 3: 800 4: 800 
Graphics: Device-1: NVIDIA GP107 [GeForce GTX 1050 Ti] vendor: eVga.com. driver: nvidia v: 435.21 bus ID: 01:00.0 
Display: x11 server: X.Org 1.20.8 driver: nvidia resolution: 1680x1050~60Hz 
OpenGL: renderer: GeForce GTX 1050 Ti/PCIe/SSE2 v: 4.6.0 NVIDIA 435.21 direct render: Yes 
Audio: Device-1: Advanced Micro Devices [AMD/ATI] SBx00 Azalia vendor: Gigabyte driver: snd_hda_intel v: kernel 
bus ID: 00:14.2 
Device-2: NVIDIA GP107GL High Definition Audio vendor: eVga.com. driver: snd_hda_intel v: kernel bus ID: 01:00.1 
Device-3: XMOS type: USB driver: hid-generic,snd-usb-audio,usbhid bus ID: 1-1:2 
Sound Server: ALSA v: k5.3.0-62-generic 
Network: Device-1: Realtek RTL8111/8168/8411 PCI Express Gigabit Ethernet vendor: Gigabyte driver: r8169 v: kernel 
port: de00 bus ID: 03:00.0 
IF: enp3s0 state: down mac: <filter> 
Device-2: Realtek RTL8111/8168/8411 PCI Express Gigabit Ethernet vendor: Gigabyte driver: r8169 v: kernel 
port: ce00 bus ID: 04:00.0  
IF: enp4s0 state: up speed: 100 Mbps duplex: full mac: <filter> 
Drives:
ID-1: /dev/sdb vendor: Crucial model: M4-CT128M4SSD2 size: 119.24 GiB 
Partition: ID-1: / size: 32.45 GiB used: 24.12 GiB (74.3%) fs: ext4 dev: /dev/sdb3 
Sensors: System Temperatures: cpu: 39.0 C mobo: N/A gpu: nvidia temp: 27 C 
Fan Speeds (RPM): N/A gpu: nvidia fan: 30% 
Info: Processes: 233 Uptime: 24m Memory: 15.64 GiB used: 1.25 GiB (8.0%) Init: systemd runlevel: 5 Compilers: gcc: 7.5.0 
Shell: bash v: 4.4.20 inxi: 3.0.32

aplay -l
**** List of PLAYBACK Hardware Devices ****
card 0: SB [HDA ATI SB], device 0: ALC889A Analog [ALC889A Analog]
Subdevices: 1/1
Subdevice #0: subdevice #0
card 0: SB [HDA ATI SB], device 1: ALC889A Digital [ALC889A Digital]
Subdevices: 1/1
Subdevice #0: subdevice #0
card 1: NVidia [HDA NVidia], device 3: HDMI 0 [HDMI 0]
Subdevices: 1/1
Subdevice #0: subdevice #0
card 1: NVidia [HDA NVidia], device 7: HDMI 1 [HDMI 1]
Subdevices: 1/1
Subdevice #0: subdevice #0
card 1: NVidia [HDA NVidia], device 8: HDMI 2 [HDMI 2]
Subdevices: 1/1
Subdevice #0: subdevice #0
card 2: Audio [iFi (by AMR) HD USB Audio], device 0: USB Audio [USB Audio]
Subdevices: 1/1
Subdevice #0: subdevice #0

**Comment by Oleksiy-Yakovenko** [comment]
Let's break down to invdidual points and questions that you posted.

> I want to be able to play PCM as flac and DSD (although that is probably a separate issue).

I don't know what this means.

> With Alsa it plays all types of music files (flac, wav, Mp3) at the wrong pitch - 16/44.1 plays a little too fast and higher res eg 24/96 very slow.

You need to restore deadbeef to default settings. The problem you described usually occurs when you remove the resampler from DSP chain, or when you misconfigure it.

> My aim is to get Deadbeef to work perfectly with Alsa to obtain a bit-perfect playback system.

That is exactly what you have now. It's bit perfect. And that is exactly why you're getting wrong speed.
If you enable resampling -- it won't be bit-perfect anymore, since the stream has to be modified to play at the right speed.
And no, you can't have both the resampler and bit-perfect playback at the same time.

**Reply by thanks6** [reply]
I have had time to digest all of your points. Thank you. I uninstalled v. 1.8.2 and installed v. 1.8.4 to restore default config. I can now see the DSP options and the resampler (SRC) seems to be the main factor. When removed, files play at the wrong speed. If “Autodetect samplerate from output device” is checked files play at the correct speed. However, you are saying that this is modifying the file. Therefore its not possible get bit-perfect playback with Deadbeef even with Alsa. Is this correct?

With regard to my DAC. It can support all the different sample rates. So I unchecked the Override samplerate option with the Resampler removed but the files play at the wrong speed. When I add the Resampler and check Autocheck samplerate…..” it plays correctly.

**Reply by Oleksiy-Yakovenko** [reply]
> Therefore its not possible get bit-perfect playback with Deadbeef even with Alsa. Is this correct?

If your DAC doesn't support the samplerate of the audio file -- resampling is required
If your DAC supports the samplerate of the audio file - resampling is not required.

You can find out which samplerate was selected on the DAC by looking in various files in /proc
Something like `/proc/asound/card0/pcm0p/sub0/hw_params`
But you'll need to figure out which file corresponds to which sound card.

**Reply by thanks6** [reply]
Hi my DAC (ifi-audio Zen Dac) supports all of the sample rates of the files I am playing and has a led that indicates the sample rate being received by the DAC. If I use Foobar 2000 in Windows the correct LED colour illuminates eg Green - PCM 44/48/88/96, Yellow - PCM 176/192. When playing 192 sample rate the LED stays green.
 The hw_params states the rate is only 48000 (48000/1). This suggests that something is limiting the  DAC. Are you able to offer any suggestions to solve this issue. Thank you.

**Reply by c3kkos** [reply]
> Let's break down to invdidual points and questions that you posted.
> 
> > I want to be able to play PCM as flac and DSD (although that is probably a separate issue).
> 
**> I don't know what this means.**
> 


Sorry to be pointy here,  Oleksiy-Yakovenko

that is a rightful question. 

I have a Fiio K3 headphone amp/dac and i'm having the same issues.

BUT.

I got issues only with DSD files playback. (slowed down and horrible pitch)
 Got no issues with 24b/192khz flacs.

IMHO The problem IS NOT outside deadbeef as stated here. <- EDIT: of course is outside. i'm wrong. its a FFMPEG decoding issue.

Although an ALSA bug is still possible, its very unlikely.

I suggest you to try out .dsf files playback with ALSA and direct hw access, of course with NO RESAMPLING.

To clean away cloudy concepts:

bit perfect = no resample, no effects, no mixing, NOTHING but the music, being passed to hw buffers.

**Comment by Oleksiy-Yakovenko** [comment]
If you have a DAC which supports all the different samplerates in hardware, you can remove the resampler, but you have to disable the "Override samplerate" option in the preferences Sound tab, so that deadbeef knows to change the DAC samplerate to match the audio file samplerate.

**Comment by thanks6** [comment]
Thank you so much for your replies. 

I want to be able to play PCM as flac and DSD (although that is probably a separate issue). I just wanted to explain that my PCM as opposed to eg DSD are encoded as flac. I also want to play DSD files but I am no sure if Deadbeef can do this but my main concern is the incorrect pitch/speed at this point.

You need to restore deadbeef to default settings. The problem you described usually occurs when you remove the resampler from DSP chain, or when you misconfigure it.

I haven’t removed the resampler. How do I restore Deadbeef - should I uninstall/reinstall?

  If you have a DAC which supports all the different samplerates in hardware, you can remove the resampler, but you have to disable the "Override samplerate" option in the preferences Sound tab, so that deadbeef knows to change the DAC samplerate to match the audio file samplerate.

Yes my DAC supports different sample rates. The “Override sample rate” is unchecked. 

Thank you for your time and effort

**Reply by pinballmachine** [reply]
Bit of a late reply, but perhaps still useful for you, or for someone else.

The problem is outside of Deadbeef. The problem is with likely with Pulseaudio. However there could be a small bug in the ALSA output plugin, see below.

Here is how I solved it:
1. create a Pulseaudio config file: ~/.config/pulse/daemon.conf
2. put in this line: `avoid-resampling = true`
3. create a ALSA config file: ~/.asoundrc
4. Put in this:
```
pcm.device{
    type hw
    card 1
    device 0
}

pcm.!default{
    type plug
    slave.pcm "device"
}
```

This sets the USB dac as the default device. It assumes that the USB dac is 'card 1 device 0'. Usually card 0 device 0 is your onboard audio device.
5. To ensure that after reboots the USB dac remains card 1 device 0 and doesn't get switched around with the onboard audio, create this file: /etc/modprobe.d/alsa-base.conf
6. Put in these two lines:
```
options snd_hda_intel index=0
options snd-usb-audio index=1
```
If your onboard audio is called differently, you need to find that out and change the first line accordingly.

In Deadbeef I use the ALSA output plugin and for 'Output device' I have selected 'Default audio device'.
Strangely, under the Plugins tab, under the ALSA output plugin, I have to keep 'Use ALSA resampling' checked. If I uncheck this, and change to a song with different samplerate, the switching doesn't work the first time. I have to restart the new song and only then the samplerate gets switched. This is possibly a bug in the ALSA output plugin.

**Comment by c3kkos** [comment]
I got a Fiio K3, playing ALL GOOD, even dsd64

got slowed down pitch with dsd128


aplay -l
**** List of PLAYBACK Hardware Devices ****
card 0: Intel [HDA Intel], device 0: 92HD73C1X5 Analog [92HD73C1X5 Analog]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 1: HDMI [HDA ATI HDMI], device 3: HDMI 0 [HDMI 0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 2: K3 [K3], device 0: USB Audio [USB Audio]
  Subdevices: 1/1
  Subdevice #0: subdevice #0

Deadbeef from .deb package v. 1.9.4
Linux Mint with liquorix kernel: 6.1.0-6.3-liquorix-amd64



as a last note:

the Fiio K3 has a led indicating BLUE for up-to-44,1khz sampling.

YELLOW for PCM above 44.1

GREEN for DSDs


well, it goes blu, it goes yellow... it never goes green.


that means dsds are converted to high sampling rate PCM maybe?

oh, i got alsa resampling disabled, of course, in the plugin tab

this confirms:

/proc/asound/card2/pcm0p/sub0$ cat hw_params 
access: RW_INTERLEAVED
format: S16_LE
subformat: STD
channels: 2
rate: 352800 (352800/1)
period_size: 1024
buffer_size: 8192


EDIT: as i wrote without having the whole picture clear, of course the LED of K3 is yellow even when deadbeef is playing DSDs

That work as intended, the decoding is a FFMPEG duty. It takes dsd data and translate to 384khz, 16bit, regular PCM.



Still,

the reason why a DSD64 playback is correct, while a DSD128 is not (like played at 0.5x speed) is not clear.
That remains for me the only issue so far.
